src/gencost/waterfall.py

```diff
@@ -2256,3 +2267,480 @@ def valid_generation(df_):
+                    0,
+                    x["ba_plus_essentials_present"],
+                ),
+                fill_in_score=lambda x: x["essentials_present"]
+                + x["ba_plus_essentials_present"]
+                + x["ba_plus_age_present"],
+            )
+        )
+
+        # some sort of for loop that sets prioritizes merging based on fill in score
+
+        three = (
+            missing_data.query("fill_in_score ==3")
+            .merge(
+                hist_data.drop(
+                    columns=[
+                        "plant_id_eia",
+                        "generator_id",
+                        "year",
+                        "prime_mover",
+                        "fuel_group",
+                        "final_ba_code",
+                        "age_in_current_year",
+                        "age_range",
+                        "essentials",
+                        "ba_plus_essentials",
+                    ]
+                ),
+                on=["ba_plus_age"],
+                how="left",
+                indicator=True,
+            )
+            .query('_merge == "both"')
+            .drop_duplicates(subset=["plant_id_eia", "generator_id", "year"])
+        )
+
+        two = (
+            missing_data.query("fill_in_score ==2")
+            .merge(
+                hist_data.drop(
+                    columns=[
+                        "plant_id_eia",
+                        "generator_id",
+                        "year",
+                        "prime_mover",
+                        "fuel_group",
+                        "final_ba_code",
+                        "age_in_current_year",
+                        "age_range",
+                        "ba_plus_age",
+                        "essentials",
+                    ]
+                ),
+                on=["ba_plus_essentials"],
+                how="left",
+                indicator=True,
+            )
+            .query('_merge == "both"')
+            .drop_duplicates(subset=["plant_id_eia", "generator_id", "year"])
+        )
+
+        one = (
+            missing_data.query("fill_in_score ==1")
+            .merge(
+                hist_data.drop(
+                    columns=[
+                        "plant_id_eia",
+                        "generator_id",
+                        "year",
+                        "prime_mover",
+                        "fuel_group",
+                        "final_ba_code",
+                        "age_in_current_year",
+                        "age_range",
+                        "ba_plus_age",
+                        "ba_plus_essentials",
+                    ]
+                ),
+                on=["essentials"],
+                how="left",
+                indicator=True,
+            )
+            .query('_merge == "both"')
+            .drop_duplicates(subset=["plant_id_eia", "generator_id", "year"])
+        )
+
+        return (
```

Comment:
From my read of this function, what we have here are the rows of missing data, to create the full eternal present dataset, we need to combine this with historical data? Or is the historical data we want to keep here already?